st2common/st2common/services/trace.py

```diff
@@ -0,0 +1,258 @@
+    if isinstance(trace_context, dict):
+        trace_context = TraceContext(**trace_context)
+
+    return trace_context
+
+
+def _get_single_trace_by_component(**component_filter):
+    """
+    Tries to return a single Trace mathing component_filter. Raises an exception
+    when a filter matches multiple.
+    """
+    traces = Trace.query(**component_filter)
+    if len(traces) == 0:
+        return None
+    elif len(traces) > 1:
+        raise UniqueTraceNotFoundException(
+            'More than 1 trace matching %s found.' % component_filter)
+    return traces[0]
+
+
+def get_trace_db_by_action_execution(action_execution=None, action_execution_id=None):
+    if action_execution:
+        action_execution_id = str(action_execution.id)
+    return _get_single_trace_by_component(action_executions__object_id=action_execution_id)
+
+
+def get_trace_db_by_rule(rule=None, rule_id=None):
+    if rule:
+        rule_id = str(rule.id)
+    # by rule could return multiple traces
+    return Trace.query(rules__object_id=rule_id)
+
+
+def get_trace_db_by_trigger_instance(trigger_instance=None, trigger_instance_id=None):
+    if trigger_instance:
+        trigger_instance_id = str(trigger_instance.id)
+    return _get_single_trace_by_component(trigger_instances__object_id=trigger_instance_id)
+
+
+def get_trace(trace_context, ignore_trace_id=False):
+    """
+    :param trace_context: context object using which a trace can be found.
+    :type trace_context: ``dict`` or ``TraceContext``
+
+    :param ignore_trace_id: Even if a trace_id is provided will be ignored.
+    :type ignore_trace_id: ``str``
+
+    :rtype: ``TraceDB``
+    """
+
+    trace_context = _get_valid_trace_context(trace_context)
+
+    if not trace_context.id_ and not trace_context.trace_id:
+        raise ValueError('Atleast one of id_ or trace_id should be specified.')
+
+    if trace_context.id_:
+        try:
+            return Trace.get_by_id(trace_context.id_)
+        except (ValidationError, ValueError):
+            LOG.warning('Database lookup for Trace with id="%s" failed.',
+                        trace_context.id_, exc_info=True)
+            raise StackStormDBObjectNotFoundError(
+                'Unable to find Trace with id="%s"' % trace_context.id_)
+
+    if ignore_trace_id:
+        return None
+
+    traces = Trace.query(trace_id=trace_context.trace_id)
+
+    # Assume this method only handles 1 trace.
+    if len(traces) > 1:
+        raise UniqueTraceNotFoundException(
+            'More than 1 Trace matching %s found.' % trace_context.trace_id)
+
+    return traces[0]
+
+
+def get_trace_db_by_live_action(liveaction):
+    """
+    Given a liveaction does the best attempt to return a TraceDB.
+    1. From trace_context in liveaction.context
+    2. From parent in liveaction.context
+    3. From action_execution associated with provided liveaction
+    4. Creates a new TraceDB (which calling method is on the hook to persist).
```

Comment:
Not a fan of this. Why can't this be done by the caller? if not get_trace_db(liveaction): create_trace()